File: SourceCode_Programmers/이분탐색/예산.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> budgets, int M) {
    int answer = 0;
    int totalSum = 0;
    int budgetsSize = budgets.size();
    
    sort(budgets.begin(), budgets.end());
    
    for(int i=0; i<budgetsSize; ++i)
    {
        totalSum += budgets[i];
    }
    
    if(totalSum <= M)
        answer = budgets[budgetsSize - 1];
    else
    {
        int start = M / budgetsSize;
        int end =  totalSum / budgetsSize;
        
        if(start > end)
        {
            int tempNum = start;
            start = end;
            end = tempNum;
        }
        
        int maxBudget = 0;
        while(true)
        {
            int tempPos = (start + end) / 2;
            int tempSum = 0;
            bool checkValue = false;
            
            
            for(int i=0; i<budgetsSize; ++i)
            {
                if(tempPos > budgets[i])
                    tempSum += budgets[i];
                else
                    tempSum += tempPos;
                
                if(tempSum > M)
                {
                    checkValue = true;
                    break;
                }
            }
            
            if(checkValue)
            {
                end = tempPos - 1;
                continue;
            }
            else
            {
                start = tempPos + 1;
            }
            
            if(maxBudget >= tempPos)
                break;
            maxBudget = tempPos;
        }
        
        answer = maxBudget;        
    }
    return answer;
}

File: SourceCode_Programmers/예산.js
function solution(d, budget) {
  let answer = 0;
  // 각 부서에 필요한 물품을 지원해 주기 위해 부서별로 물품을 구매하는데 필요한 금액 조사
  // 전체 예산이 정해져 있음
  // 최대한 많은 부서의 물품을 구매해 줄 수 있도록 하려고 함
  // 딱 맞게 금액을 지원함
  // d, budget
  
  const sortDList = d.sort((itemA, itemB) => itemA - itemB);
  const dLen = d.length;
  let totalSum = 0;
  for (; answer < dLen; answer += 1) {
      totalSum += sortDList[answer];
      
      if (totalSum > budget) {
          break;
      }
  }
  return answer;
}